usr/lib/common/key.c

```diff
@@ -1660,6 +1663,133 @@ CK_RV rsa_priv_unwrap_get_data(TEMPLATE *tmpl,
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_TR, &tr) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_TR for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S1, &s1) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_S1 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_S2, &s2) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_S2 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T0, &t0) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_T0 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+
+        if (template_attribute_find(tmpl, CKA_IBM_DILITHIUM_T1, &t1) == FALSE) {
+            TRACE_ERROR("Could not find CKA_IBM_DILITHIUM_T1 for the key.\n");
+            return CKR_FUNCTION_FAILED;
+        }
+    }
+
+    rc = ber_encode_IBM_DilithiumPrivateKey(length_only, data, data_len,
+                                  rho, seed, tr, s1, s2, t0, t1,
+                                  opaque);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_encode_IBM_DilithiumPrivateKey failed\n");
+    }
+
+    return rc;
+}
+
+CK_RV ibm_dilithium_priv_unwrap_get_data(TEMPLATE *tmpl, CK_BYTE *data,
+                                         CK_ULONG total_length)
+{
+    CK_ATTRIBUTE *rho = NULL;
+    CK_ATTRIBUTE *t1 = NULL;
+    CK_RV rc;
+
+    rc = ber_decode_IBM_DilithiumPublicKey(data, total_length, &rho, &t1);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("ber_decode_DilithiumPublicKey failed\n");
+        return rc;
+    }
+
+    rc = template_update_attribute(tmpl, rho);
+    if (rc != CKR_OK)
+        TRACE_DEVEL("template_update_attribute(CKA_IBM_DILITHIUM_RHO) failed\n");
+    rc = template_update_attribute(tmpl, t1);
+    if (rc != CKR_OK)
+        TRACE_DEVEL("template_update_attribute(CKA_IBM_DILITHIUM_T1) failed\n");
+
+    return CKR_OK;
+}
+
+//
+//
+CK_RV ibm_dilithium_priv_unwrap(TEMPLATE *tmpl, CK_BYTE *data,
+                                CK_ULONG total_length, CK_BBOOL isOpaque)
+{
+    CK_ATTRIBUTE *rho = NULL, *seed = NULL, *tr = NULL;
+    CK_ATTRIBUTE *s1 = NULL, *s2 = NULL, *t0 = NULL, *t1 = NULL;
+    CK_ATTRIBUTE *opaque = NULL;
+    CK_RV rc;
+
+    rc = ber_decode_IBM_DilithiumPrivateKey(data, total_length,
+                                 &rho, &seed, &tr, &s1, &s2, &t0, &t1,
+                                 &opaque, isOpaque);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("der_decode_IBM_DilithiumPrivateKey failed\n");
+        return rc;
+    }
+
+    if (isOpaque)
+        template_update_attribute(tmpl, opaque);
+
+    template_update_attribute(tmpl, rho);
```

Comment:
Error checking ?  Maybe like rc |= template_update_attribute ... and check rc at the end?

Reply:
Yes, agreed.
Also here we have many routines coded this way, so we should do some overall cleanup too.

Reply:
While changing this code, I realized that we use TRACE_DEVEL if any ber_decode_xxx or template_update_attribute routine failed. Shouldn't we better use TRACE_ERROR here? TRACE_DEVEL may be additional output of key parts, but here we really have error situations. What do you think?